### PR TITLE
During sign up only allow email address the invite was send to

### PR DESCRIPTION
**Pull request by MartinRiese**
## Product Description

Add setting to projects "privacy and security" page that allows restricting the email address used for sign up to a project

image

If enabled the setting will result in the email field being disabled in the form and the code handling the submission checking the email address.

image

## Technical Summary

https://dimagi-dev.atlassian.net/browse/USH-2578

## Feature Flag

None

### Safety story

The code changes are additive in nature making the risk fairly low.

### Automated test coverage

Updated test checking visible security settings

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/users/views/web.py [hunk 1]
```diff
@@ -158,6 +164,12 @@ def __call__(self, request, uuid, **kwargs):
                         signup_request = AsyncSignupRequest.create_from_invitation(invitation)
                         return HttpResponseRedirect(idp.get_login_url(signup_request.username))
 
+                    if allow_invite_email_only and \
+                            request.POST.get("email").lower() != invitation.email.lower():
+                        messages.error(request, _("You can only sign up with the email "
+                                                  "address your invitation was send to."))
```
Comment (Robert-Costello) [inline corehq/apps/users/views/web.py:170]: Small spelling/grammar fix: "send" should be "sent"

File: corehq/apps/domain/forms.py [hunk 2]
```diff
@@ -804,6 +804,11 @@ class PrivacySecurityForm(forms.Form):
         help_text=gettext_lazy("Mobile Workers will never be locked out of their account, regardless"
             "of the number of failed attempts")
     )
+    allow_invite_email_only = BooleanField(
+        label=gettext_lazy("During sign up only allow the email address the invite was send to"),
```
Comment (orangejenny) [inline corehq/apps/domain/forms.py:808]:
```suggestion
        label=gettext_lazy("During sign up, only allow the email address the invitation was sent to"),
```

**Review by Robert-Costello — APPROVED** [review]

**Review by nospame — COMMENTED** [review]
Looks good to me. Is there any way to add a test for this? Something like, with the option enabled, submit the `WebUserInvitationForm` with the wrong email and make sure it throws an error.

**Comment by MartinRiese** [comment]
@nospame Do the tests look good to you?

File: corehq/apps/users/tests/test_user_invitation.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        invite_uuid = "e1bd37f5-9ff8-4853-b953-fd75483a0ec7"
+        domain = "domain"
+
+        invite = Invitation.objects.create(
+            uuid=invite_uuid,
+            domain=domain,
+            is_accepted=True,
+            invited_on=datetime.date(2023, 9, 1)
+        )
+        invite.save()
```
Comment (millerdev) [inline corehq/apps/users/tests/test_user_invitation.py:46]: nit: unneeded line. `.create(...)` does a db insert.

File: corehq/apps/users/tests/test_user_invitation.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        form = StubbedWebUserInvitationForm(
+            {
+                "email": "other_test@dimagi.com",
+                "full_name": "asdf",
+                "password": "pass",
+            },
+            is_sso=False,
+            allow_invite_email_only=True,
+            invite_email="test@dimagi.com",
+            request_email="other_test@dimagi.com",)
```
Comment (millerdev) [inline corehq/apps/users/tests/test_user_invitation.py:62]: nit: typical wrapping in this style would put the closing paren on a new line and dedented to the level of the line with the opening paren.
```suggestion
            request_email="other_test@dimagi.com",
        )
```
There are a few other places in this PR where this comment also applies.

**Review by millerdev — COMMENTED** [review]
Random review because the PR title looked interesting. Not going to approve since I don't have full context, but left a couple nitpicky comments.

File: corehq/apps/users/tests/test_user_invitation.py (hunk 3 above)
Comment (MartinRiese) [reply corehq/apps/users/tests/test_user_invitation.py:46]: Good to know. Removed

**Pull request merged**